things/postgres/groups.go

```diff
@@ -0,0 +1,626 @@
+	qd := `DELETE FROM thing_groups WHERE id = :id`
+	group := groups.Group{
+		ID: groupID,
+	}
+	dbg, err := toDBGroup(group)
+	if err != nil {
+		return errors.Wrap(errUpdateDB, err)
+	}
+
+	res, err := gr.db.NamedExecContext(ctx, qd, dbg)
+	if err != nil {
+		return errors.Wrap(errDeleteGroupDB, err)
+	}
+
+	cnt, err := res.RowsAffected()
+	if err != nil {
+		return errors.Wrap(errDeleteGroupDB, err)
+	}
+
+	if cnt != 1 {
+		return errors.Wrap(groups.ErrDeleteGroup, err)
+	}
+	return nil
+}
+
+func (gr groupRepository) RetrieveByID(ctx context.Context, id string) (groups.Group, error) {
+	groupID, err := toUUID(id)
+	if err != nil || !groupID.Valid {
+		return groups.Group{}, err
+	}
+
+	dbu := dbGroup{
+		ID: groupID,
+	}
+
+	q := `SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups WHERE id = $1`
+	if err := gr.db.QueryRowxContext(ctx, q, id).StructScan(&dbu); err != nil {
+		if err == sql.ErrNoRows {
+			return groups.Group{}, errors.Wrap(groups.ErrNotFound, err)
+
+		}
+		return groups.Group{}, errors.Wrap(errRetrieveDB, err)
+	}
+
+	return toGroup(dbu)
+}
+
+func (gr groupRepository) RetrieveAll(ctx context.Context, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	_, mq, err := getGroupsMetadataQuery("thing_groups", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+
+	q := fmt.Sprintf(`SELECT id, owner_id, parent_id, name, description, metadata, path, nlevel(path) as level FROM thing_groups 
+					  WHERE nlevel(path) <= :level %s ORDER BY path`, mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE nlevel(path) <= :level %s", mq)
+
+	dbPage, err := toDBGroupPage("", "", "", "", level, gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	items, err := processRows(rows)
+	if err != nil {
+		return groups.GroupPage{}, err
+	}
+
+	total, err := total(ctx, gr.db, cq, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := groups.GroupPage{
+		Groups: items,
+		PageMetadata: groups.PageMetadata{
+			Total: total,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) RetrieveAllParents(ctx context.Context, groupID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	if groupID == "" {
+		return groups.GroupPage{}, nil
+	}
+
+	_, mq, err := getGroupsMetadataQuery("thing_groups", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+
+	q := fmt.Sprintf(`SELECT g.id, g.name, g.owner_id, g.parent_id, g.description, g.metadata, g.path, nlevel(g.path) as level FROM 
```

Comment:
Consider using simpler `self join`. You can try this query:
```sql
SELECT g.id, g.name, g.owner_id, g.parent_id, g.description, g.metadata, g.path, nlevel(g.path) as level
FROM thing_groups parent, thing_groups g
WHERE parent.id = :id AND g.path @> parent.path AND nlevel(parent.path) - nlevel(g.path) <= :level;
```
It's a little simpler (and shorter) and possibly more performant.